### PR TITLE
Implement ingest & load for the ETH canopy height dataset

### PR DESCRIPTION
**Pull request by BSchilperoort**
In this PR the Zampy Dataset protocol is fleshed out, and implemented for the ETH canopy height dataset. 

`eth_dataset_demo.ipynb` contains a demonstration of the functionality.

### CODE REVIEW

File: src/zampy/datasets/validation.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            f"\nThe '{dataset.name}' data could not cover the start of requested range:"
+            f"\n    data start: {dataset.time_bounds.start}"
+            f"\n    requested start: {request_bounds.start}"
+        )
+    if dataset.time_bounds.end < request_bounds.end:
+        error_message += (
+            f"\nThe '{dataset.name}' data could not cover the end of requested range:"
+            f"\n    data end: {dataset.time_bounds.end}"
+            f"\n    requested end: {request_bounds.end}"
+        )
```
Comment (geek-yang) [inline src/zampy/datasets/validation.py:87]: If I configure my time bound as `times = TimeBounds(np.datetime64("2022-01-01"), np.datetime64("2019-12-31"))`, the downloading could still run.

We need to add another check to make sure that `request_bounds.start < request_bounds.end`.

**Review by geek-yang — CHANGES_REQUESTED** [review]
Very nice work! This serves as a very good starting point for the following development, which makes adding new functionality and support for other datasets much easier. I really like the design for enclosing all the docstrings in the "template" code, making the dataset script quite clean and elegant.

I have some comments regarding the check. When reviewing the code, I simply assume that unit tests and other things related to the best practices will be added later (in another PR), am I correct? since this PR always grows quite bigger. Most of the checks could be added later, but at least we can leave a TODO in the code, as there are some checks missing. Please see my comments for the details.

Overall, it looks pretty awesome! Thanks a lot for the architecture-level design 😎.

File: src/zampy/datasets/validation.py (hunk 1 above)
Comment (BSchilperoort) [reply src/zampy/datasets/validation.py:87]: I think this validator should be implemented in the TimeBounds class: the end time should be later than the start time.

File: src/zampy/datasets/eth_canopy_height.py [hunk 2]
```diff
@@ -70,16 +82,62 @@ def download(  # noqa: PLR0913
...
+
+        for file, sd_file in zip(data_files + sd_files, is_sd_file):
+            convert_tiff_to_netcdf(
+                ingest_folder,
+                file=file,
+                sd_file=sd_file,
+                overwrite=overwrite,
+            )
+
+        copy_properties_file(download_folder, ingest_folder)
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply src/zampy/datasets/eth_canopy_height.py:117]: Well, the directories should probably be different. However, the code that should manage that needs to be in the recipe parsing part.

The functionality of this PR is not designed to be end-user accessible: the end user will interact through recipes and a CLI.
Not having to add all these checks here will make it easier and cleaner to implement the datasets.

File: src/zampy/datasets/eth_canopy_height.py [hunk 3]
```diff
@@ -70,16 +82,62 @@ def download(  # noqa: PLR0913
                 fpath=download_folder / fname,
                 overwrite=overwrite,
             )
+
+        write_properties_file(
+            download_folder, spatial_bounds, time_bounds, variable_names
+        )
+
         return True
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply src/zampy/datasets/eth_canopy_height.py:90]: The return code is more for the rest of our code which will interact with the Dataset's methods.

User interaction is then done through a unified interface, without needing specific implementations in each dataset.

File: src/zampy/datasets/eth_canopy_height.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (geek-yang) [reply src/zampy/datasets/eth_canopy_height.py:117]: That's clear to me then. So in the recipe `example_recipe`, the user only needs to specify the `folder` and zampy will take care of the rest, right? Then those checks may not necessary, I agree.

**Comment by BSchilperoort** [comment]
Thanks for the review, Yang! Unit tests etc. will indeed be added later. This PR is mostly on getting a working design, and I did not want to implement tests that I would need to rewrite anyway.

However, I should add the tests soon and not wait too long after merging this :sweat_smile:

**Pull request merged**